File: moonpoison/binary_tree-dfs/104.cpp
#include<bits/stdc++.h>
using namespace std;
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
class Solution {
public:
    int MAX=0;
    int maxDepth(TreeNode* root)
    {
        if(root==NULL) return NULL;
        dfs(root, 1);
        return MAX;
    }
    void dfs(TreeNode* root, int cnt)
    {
        if(root->left!=nullptr) dfs(root->left, cnt+1);
        if(root->right!=nullptr) dfs(root->right, cnt+1);
        if(root->left==nullptr&&root->right==nullptr) MAX=max(MAX, cnt);
    }
};

File: ice0208/binary-tree-dfs/104.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def down(self, cursor, depth):
        self.max_depth = max(self.max_depth, depth)

        if cursor.left:
            self.down(cursor.left, depth + 1)
        if cursor.right:
            self.down(cursor.right, depth + 1)

    def maxDepth(self, root) -> int:
        self.max_depth = 0

        if root:
            self.down(root, 1)

        return self.max_depth
